v-next/hardhat/test/internal/builtin-plugins/network-manager/hook-handlers/config.ts

```diff
@@ -0,0 +1,669 @@
+      const configWithInvalidGasPrice = {
+        networks: {
+          localhost: {
+            type: "http",
+            url: "http://localhost:8545",
+            gasPrice: "invalid",
+          },
+        },
+      };
+
+      let validationErrors = await validateUserConfig(
+        /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+        -- testing invalid network type for js users */
+        configWithInvalidGasPrice as any,
+      );
+
+      assert.ok(
+        validationErrors.length > 0,
+        "validation errors should be present",
+      );
+      assert.equal(
+        validationErrors[0].message,
+        "Expected 'auto', a safe int, or bigint",
+      );
+
+      const configWithNonSafeIntGasPrice = {
+        networks: {
+          localhost: {
+            type: "http",
+            url: "http://localhost:8545",
+            gasPrice: Number.MAX_SAFE_INTEGER + 1,
+          },
+        },
+      };
+
+      validationErrors = await validateUserConfig(
+        /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+        -- testing invalid network type for js users */
+        configWithNonSafeIntGasPrice as any,
+      );
+
+      assert.ok(
+        validationErrors.length > 0,
+        "validation errors should be present",
+      );
+
+      // TODO: the error message should be "Expected 'auto', a safe int, or bigint"
+      assert.equal(
+        validationErrors[0].message,
+        "Number must be less than or equal to 9007199254740991",
+      );
+    });
+
+    describe("http network specific fields", () => {
+      it("should throw if the url is missing", async () => {
+        const config = {
+          networks: {
+            localhost: {
+              type: "http",
+            },
+          },
+        };
+
+        /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+        -- testing invalid network type for js users */
+        const validationErrors = await validateUserConfig(config as any);
+
+        assert.ok(
+          validationErrors.length > 0,
+          "validation errors should be present",
+        );
+        assert.equal(
+          validationErrors[0].message,
+          "Expected a URL or a Configuration Variable",
+        );
+      });
+
+      it("should throw if the url is invalid", async () => {
+        const config = {
+          networks: {
+            localhost: {
+              type: "http",
+              url: "invalid",
+            },
+          },
+        };
+
+        /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+        -- testing invalid network type for js users */
+        const validationErrors = await validateUserConfig(config as any);
+
+        assert.ok(
+          validationErrors.length > 0,
+          "validation errors should be present",
+        );
+        // TODO: the error message should be "Expected a URL or a Configuration Variable"
```

Comment:
The interesting thing is that this validation uses `conditionalUnionType` behind the scenes, so it might be bugged as well.

Reply:
I think this is working as intended concerning the `conditionalUnionType`. Its main idea is to return errors that are as specific as possible.

`unionType` does the contrary. It always returns a less specific error message.

Maybe for this particular case we should use `unionType`.

Reply:
the problem is that `unionType` expects the first argument to be an array of zod types, and we're using `isObject` in `sensitiveUrlType`. Maybe we could replace it with `z.object()` or `z.record()`, but we'd need to make it generic.

Reply:
I've updated `sensitiveUrlType` and `sensitiveStringType` to use `unionType` (and renamed them to schemas). However, there's a bug in Zod that prevents the union from using the provided `errorMap` when the string schema is `z.string().url()` (and likely other validations). See: https://github.com/NomicFoundation/hardhat/blob/ae009033ab5cb130042b33707f17a1ca12631f06/v-next/hardhat-zod-utils/src/index.ts#L156-L160